### PR TITLE
Iterator based compilation

### PR DESCRIPTION
**Pull request by Schaeff**
Todo:
- [x] design `inspect` command
- [x] implement unconstrained detection
- [x] implement stream deserialisation -> will happen with #1045

### CODE REVIEW

File: zokrates_core/src/ir/serialize.rs [hunk 1]
```diff
@@ -1,30 +1,121 @@
...
+    out: &mut W,
+    prog_iterator: ProgIterator<T, I>,
+) -> serde_cbor::Result<()>
+where
+    I: IntoIterator<Item = Statement<T>>,
+    W: Write,
+{
+    struct Wrapper<U>(Cell<Option<U>>);
+
+    struct SerializableProgIterator<T: Serialize, I: IntoIterator<Item = crate::ir::Statement<T>>> {
```
Comment (dark64) [inline zokrates_core/src/ir/serialize.rs]: `Statement` is already imported so `crate::ir::` is not needed

File: zokrates_core/src/ir/serialize.rs [hunk 2]
```diff
@@ -1,30 +1,121 @@
...
+        arguments: Vec<crate::flat_absy::FlatParameter>,
+        return_count: usize,
+        statements: Wrapper<I>,
+    }
+
+    impl<T, I> Serialize for SerializableProgIterator<T, I>
+    where
+        T: Serialize + std::fmt::Debug,
+        I: IntoIterator<Item = crate::ir::Statement<T>>,
```
Comment (dark64) [inline zokrates_core/src/ir/serialize.rs]: same as above

File: zokrates_core/src/compile.rs [hunk 3]
```diff
@@ -25,19 +24,26 @@ use zokrates_field::Field;
 use zokrates_pest_ast as pest;
 
 #[derive(Debug)]
-pub struct CompilationArtifacts<T: Field> {
-    prog: ir::Prog<T>,
-    abi: Abi,
+pub struct CompilationArtifacts<T, I: IntoIterator<Item = ir::Statement<T>>> {
+    pub prog: ir::ProgIterator<T, I>,
+    pub abi: Abi,
```
Comment (dark64) [inline zokrates_core/src/compile.rs]: is there a reason for this to be public? `abi` function (L37) is implemented for this struct which returns a reference

**Review by dark64 — COMMENTED** [review]
Nice one :rocket:, spotted some minor things

File: zokrates_core/src/compile.rs (hunk 3 above)
Comment (Schaeff) [reply zokrates_core/src/compile.rs]: Fixed with `into_inner`, the existing accessors were not sufficient because of borrowing rules.

**Review by dark64 — APPROVED** [review]
:rocket:

**Comment by TakodaS** [comment]
Why was unconstrained detection removed here?

**Comment by dark64** [comment]
> Why was unconstrained detection removed here?

I believe it was implemented back in a different pull request (commit https://github.com/Zokrates/ZoKrates/commit/64f99d75fbe15e0975fc847c61ec07a1988177b5)

**Comment by TakodaS** [comment]
> > Why was unconstrained detection removed here?
> 
> I believe it was implemented back in a different pull request (commit [64f99d7](https://github.com/Zokrates/ZoKrates/commit/64f99d75fbe15e0975fc847c61ec07a1988177b5))

The compiler option to allow unconstrained inputs no longer appears in the cli for zokrates v 0.8.3 as far as I can see.

**Pull request merged**